Feat: AVD dev containers (#3179)
aristanetworks/ansible-avd: ansible_collections/arista/avd/docs/containers/overview.md

<!--
  ~ Copyright (c) 2023 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# AVD Dev Containers Overview

> > WARNING: AVD Dev Container are currently in the early preview phase and not officially supported for any production cases.

Reviewer: ```suggestion
> WARNING: AVD Dev Containers are currently in the early preview phase and not officially supported for any production cases.
```

> > They can be still very useful for development and testing. Please report any issues and optimization suggestions.

Reviewer: ```suggestion
> They can still be very useful for development and testing. Please report any issues and optimization suggestions.
```

> > They can still be very useful for development and testing. Please report any issues and optimization suggestions.

Reviewer 2: not sure if this renders well on mkdocs

we use admonittion syntax `!!!` (and make github sad, yes)

Author: this is rendered very well by mkdocs. At least I can't spot any problems.

Reviewer 3: changed in my commit. Resolving this.

## Why Dev Containers

> Building a functional Ansible environment can be a challenging task. A manually created Ansible environment can be broken due to number of reasons:

Reviewer: So we mention the benefit of isolating the Ansible environment on the local machine from the environment where development is happening as these might have different requirements etc?

Author: done

- Problems with Python and Ansible versions
- Broken dependencies
- Interpreter path issues
- Inconsistent Ansible configuration files
- etc.

> Containers can provide a simple and portable way to build a functional, portable and isolated Ansible environment. [Dev Containers](https://containers.dev) bring that to the next level by simplifying container build process and extending the functionality. The [Dev Container Specification](https://github.com/devcontainers/spec) was started by Microsoft and has strong community support. Essentially Dev Containers are powered by following:

Reviewer 2: ```suggestion
Containers can provide a simple and portable way to build a functional and isolated Ansible environment. [Dev Containers](https://containers.dev) bring that to the next level by simplifying container build process and extending the functionality. The [Dev Container Specification](https://github.com/devcontainers/spec) was started by Microsoft and has strong community support. Essentially Dev Containers are powered by following:
```

Author: why not portable?

Reviewer 3: it was repeated in the sentence. I have reworded in my commit so resolving this.

- [Prebuilt images](https://github.com/devcontainers/images)
- [Features](https://containers.dev/features)

Dev containers can be run locally or remotely, in a private or public cloud, in a variety of [supporting tools and editors](https://containers.dev/supporting).

> ## AVD Dev Container Variants

Reviewer: Should we have an explanation of the naming convention? Like

`ghcr.io/aristanetworks/ansible-avd/<variant>:python<version>-avd-<avd version>`

Author: Certainly at a later point. Now right now, as a lot of changes expected at the start. Naming convention is not yet fixed.

Reviewer: Understood.

The AVD repository provides a number of pre-build dev container images for various use cases:

> - **base** - image that is used to build other variants. It is not intended for a direct use.

Reviewer: ```suggestion
- **base** - image that is used to build other variants. It is not intended for direct use.
```

> - **dev** - an image created for AVD contributors and to support quick testing of AVD branches.

Reviewer: Should we add a note that this container does not have the AVD collection installed?

Author: done

> - **universal** - an image with pre-installed AVD collection that can be used to build and deploy AVD configuration using a specific AVD version.

Reviewer: ```suggestion
- **universal** - an image pre-installed with AVD & CVP collections and all dependencies and python packages. This can be used to build and deploy AVD configuration using a specific AVD version.
```

Author: I'd prefer the old version here. Yes, arista.cvp is AVD requirement, but we only install AVD collection directly. Also I think it's "an image with". =)

Reviewer: Aren't dependencies installed as well? maybe we can just mention that AVD collection & dependencies are installed?

Author: Added a short comment. There will be a detailed doc later explaining the setup. But not in the initial PR.

> - **universal** - an image with pre-installed AVD collection and it's dependencies that can be used to build and deploy AVD configuration using a specific AVD version.

Reviewer 2: ```suggestion
- **universal** - an image with pre-installed AVD collection and its dependencies that can be used to build and deploy AVD configuration using a specific AVD version.
```

All images are based on [Python Slim](https://hub.docker.com/_/python) with some additional features and extensions pre-installed.

## How to Use Dev Containers

Currently the only supported option is using [VSCode](https://code.visualstudio.com/) to start a dev container.

To start using a dev container:

> 

Reviewer: Install Docker Desktop? Or is there support for other solutions?

Author: No need for the overview. I expect first users to know that. Let's keep it simple. We can add some guides for beginners later once it's fully released and if we have capacity to write that.

Reviewer: understood.

> To start using a dev container:

Reviewer 3: ... for running AVD.

Reviewer 3: ```suggestion
To start using the containers for running AVD:
```

- Install [VScode Dev Containers extension](https://code.visualstudio.com/docs/devcontainers/tutorial)

> - Add `.devcontainer/devcontainer.json` to your AVD repository (check examples below for possible dev container definitions).

Reviewer 3: This will already be part of the AVD repo, so maybe explain if it is another repository you talk about, like a "AVD inventory repository" or "repository containing an Ansible inventory to be used with AVD".

Reviewer 3: ```suggestion
- Add `.devcontainer/devcontainer.json` to your repository containing the AVD inventory (check examples below for possible dev container definitions).
```

- Open VSCode command pallette and pick `Dev Containers: Rebuild and Reopen in Container`
- Wait until the build will finish and enjoy running AVD in a container.

> If you want to use specific AVD release, use following `devcontainer.json`:

Reviewer: ```suggestion
If you want to use a specific AVD release, use the following `devcontainer.json`:
```

```json
{
    "name": "AVD Universal",
    "image": "ghcr.io/aristanetworks/ansible-avd/universal:python3.11-avd-v4.4.0"
}
```

> For AVD contributors following configuration is recommended:

Reviewer: ```suggestion
For AVD contributors the following configuration is recommended:
```

```json
{
    "name": "AVD Development",

>     "image": "ghcr.io/aristanetworks/ansible-avd/dev:python3.11-avd-devel",

Reviewer: typo in name

> ```json

Reviewer 3: This snip can be included from the file directly using the `--8<--` tags. Look at our role docs for examples. This will keep the doc up-to-date and maintenance free.

    "mounts": [

>         "source=${localWorkspaceFolder}/ansible_collections/arista/avd,target=/home/avd/.ansible/collections/ansible_collections/arista/avd,type=bind"

Reviewer 3: So this is confusing to me. A developer would use the development container as part of the development workflow itself. But here you are using it for *running*  AVD with the local checked out git branch. That can work fine, but it will spin up two seperate container instances right?

    ],
    // run entrypoint script manually as it's ignored by dev container CLI otherwise
    // the dev entrypoint is used to install ansible collections and requirements, as they are not included with the dev version
    // true is required to exit "onCreateCommand" without entering ZSH
    "onCreateCommand": "/bin/entrypoint.sh true"
}
```

> If you want to test specific AVD branch or fork without making any changes, use following `devcontainer.json`:

Reviewer: ```suggestion
If you want to test specific AVD branch or fork without making any changes, use the following `devcontainer.json`:
```

```json
{
    "name": "AVD Development",

>     "image": "ghcr.io/aristanetworks/ansible-avd/dev:python3.11-avd-devel",

Reviewer: typo in name

    "containerEnv": {
        "AVD_GITHUB_REPO": "aristanetworks/ansible-avd",

>         "AVD_BRANCH_NAME": "devel"

Reviewer 2: I guess you should mention in the doc that these must be changed to the appropriate variables, maybe use some syntax that shows where you need to edit like `<CHANGEME>`

Author: This is mentioned in the note below, otherwise I'd prefer a working example that can be copy-pasted. The only side effect can be that people will end up with an incorrect but working AVD version in this case. There will be more documents later explaining how to work with that.

Reviewer 2: in that case please state explicitly in the note the parameters.

```
Do not forget to adjust the parameters like AVD version according to your use case.
```

this is not obvious from the snippet how to adjust the AVD version.

    },
    // run entrypoint script manually as it's ignored by dev container CLI otherwise
    // the dev entrypoint is used to install ansible collections and requirements, as they are not included with the dev version
    // true is required to exit "onCreateCommand" without entering ZSH
    "onCreateCommand": "/bin/entrypoint.sh true"
}
```

> > NOTE: Do not forget to adjust the parameters like AVD version according to you use case.

Reviewer: ```suggestion
> NOTE: Do not forget to adjust the parameters like AVD version according to your use case.
```

- **universal** - an image with pre-installed AVD collection and its dependencies that can be used to build and deploy AVD configuration using a specific AVD version.

All images are based on [Python Slim](https://hub.docker.com/_/python) with some additional features and extensions pre-installed.

## How to Use Dev Containers

Currently the only supported option is using [VSCode](https://code.visualstudio.com/) to start a dev container.

To start using the containers for running AVD:

- Install [VScode Dev Containers extension](https://code.visualstudio.com/docs/devcontainers/tutorial)
- Add `.devcontainer/devcontainer.json` to your repository containing the AVD inventory (check examples below for possible dev container definitions).
- Open VSCode command pallette and pick `Dev Containers: Rebuild and Reopen in Container`.
- Wait until the build finish and enjoy running AVD in a container.

If you want to use a specific AVD release, use the following `devcontainer.json`:

```json
{
    "name": "AVD Universal",
    "image": "ghcr.io/aristanetworks/ansible-avd/universal:python3.11-avd-v4.4.0"
}
```

> For AVD contributors the following configuration is recommended:
> 
> ```json
> {
>     "name": "AVD Development",
>     "image": "ghcr.io/aristanetworks/ansible-avd/dev:python3.11",
>     "mounts": [
>         "source=${localWorkspaceFolder}/ansible_collections/arista/avd,target=/home/avd/.ansible/collections/ansible_collections/arista/avd,type=bind"
>     ],
>     // Run entrypoint script manually as it's ignored by dev container CLI otherwise.
>     // The dev entrypoint is used to install ansible collections and requirements, as they are not included with the dev version.
>     // "true" is required to exit "onCreateCommand" without entering ZSH.
>     "onCreateCommand": "/bin/entrypoint.sh true"
> }
> ```

Reviewer 3: ```suggestion
For AVD contributors the dev container is part of the AVD repository.
```
